Integrate transcription in PeerTube

Chocobozzz opened the pull request:
Changes in your packages:
 * Force optional options -> required options to increase constraint and force implementer (server, runner) to specify important options (logger, default transcoding
 directory...)
 * Removed `...Sync` methods in favour of async ones
 * Use objects for options instead of list of args (easier to add other options, better understand argument purpose)
 * Prefer using `fs-extra` functions instead of `fs` for consistency (`move` instead of `rename` for example that can have issues when moving a file between different devices etc.)
 * Use a type for the engine name
 * Deleted unused methods and `whisper-timestamped`
 * Added the ability to install whisper engines *on the fly* so PeerTube instance admins don't have to install them manually
 * Added a custom engine path arg so the PeerTube server can install whisper engines locally and specify their path instead of relying on the global path
 * Put custom models to download in `fixtures` directory (in `gitignore` but we have a CI cache) so we don't have to download them manually each time we run the tests
 * Reduces fixture sizes for the transcription tests (just decreased video quality while keeping the audio stream as is)
 * Deleted the transcript directory option from the constructor in favour of the method so we can instantiate only once the transcriber but use multiple times the `transcribe` method using a different directory each time
 * Created a `transcription-devtools` that includes the benchmark, jiwer and test tools, so packages that just use the transcription don't include these files
 * Removed `requirements.txt` file but added the `pip install` command in the test documentation

Added:
 * PIP/Hugging Face models cache in Github action
 * Transcription support in PeerTube runner (uses whisper engines installed globally)
 * Transcription support in PeerTube server (uses whisper engines installed on the fly in the `storage/bin/pip` directory)
  * Config to enable/disable video transcription
  * Notification for video owner when the transcription is finished
  * Display auto-transcription info in upload/import page and "features found on this instance" in about page
  * Add ability to select the auto engine/model, but admins can also specify a custom engine and model paths
 * Server and runner transcription tests

lutangar committed:
feat(transcription): groundwork

chore: fiddling around some more

chore: add ctranslate2 and timestamped

chore: add performance markers

chore: refactor test

chore: change worflow name

chore: ensure Python3

chore(duration): convert to chai/mocha syntahx

chore(transcription): add individual tests for others transcribers

chore(transcription): implement formats test of all implementations

Also compare result of other implementation to the reference implementation

chore(transcription): add more test case with other language and models size and local model

chore(test): wip ctranslate 2 adapat

chore(transcription): wip transcript file and benchmark

chore(test): clean a bit

chore(test): clean a bit

chore(test): refacto timestamed spec

chore(test): update workflow

chore(test): fix glob expansion with sh

chore(test): extract some hw info

chore(test): fix async tests

chore(benchmark): add model info

feat(transcription): allow use of a local mode in timestamped-whisper

feat(transcription): extract run and profiling info in own value object

feat(transcription): extract run concept in own class an run more bench

chore(transcription): somplify run object only a uuid is now needed and add more benchmark scenario

docs(transcription): creates own package readme

docs(transcription): add local model usage

docs(transcription): update README

fix(transcription): use fr video for better comparison

chore(transcription): make openai comparison passed

docs(timestamped): clea

chore(transcription): change transcribers transcribe method signature

Introduce whisper builtin model.

fix(transcription): activate language detection

Forbid transcript creation without a language.
Add `languageDetection` flag to an engine and some assertions.

Fix an issue in `whisper-ctranslate2` :
https://github.com/Softcatala/whisper-ctranslate2/pull/93

chore(transcription): use PeerTube time helpers instead of custom ones

Update existing time function to output an integer number of seconds and add a ms human-readable time formatter with hints of tests.

chore(transcription): use PeerTube UUID helpers

chore(transcription): enable CER evaluation

Thanks to this recent fix in Jiwer <3
https://github.com/jitsi/jiwer/issues/873

chore(jiwer): creates JiWer package

I'm not very happy with the TranscriptFileEvaluator constructor... suggestions ?

chore(JiWer): add usage in README

docs(jiwer): update JiWer readme

chore(transcription): use FunMOOC video in fixtures

chore(transcription): add proper english video fixture

chore(transcription): use os tmp directory where relevant

chore(transcription): fix jiwer cli test reference.txt

chore(transcription): move benchmark out of tests

chore(transcription): remove transcription workflow

docs(transcription): add benchmark info

fix(transcription): use ms precision in other transcribers

chore(transcription): simplify most of the tests

chore(transcription): remove slashes when building path with join

chore(transcription): make fromPath method async

chore(transcription): assert path to model is a directory for CTranslate2 transcriber

chore(transcription): ctranslate2 assertion

chore(transcription): ctranslate2 assertion

chore(transcription): add preinstall script for Python dependencies

chore(transcription): add download and unzip utils functions

chore(transcription): add download and unzip utils functions

chore(transcription): download & unzip models fixtures

chore(transcription): zip

chore(transcription): raise download file test timeout

chore(transcription): simplify download file test

chore(transcription): add transcriptions test to CI

chore(transcription): raise test preconditions timeout

chore(transcription): run preinstall scripts before running ci

chore(transcription): create dedicated tmp folder for transcriber tests

chore(transcription): raise timeout some more

chore(transcription): raise timeout some more

chore(transcription): raise timeout some more

chore(transcription): raise timeout some more

chore(transcription): raise timeout some more

chore(transcription): raise timeout some more

chore(transcription): raise timeout some more

chore(transcription): raise timeout some more

chore(transcription): use short video for local model test

chore(transcription): raise timeout some more

chore(transcription): raise timeout some more

chore(transcription): raise timeout some more

chore(transcription): setup verbosity based on NODE_ENV value

Commit:
Translated using Weblate (Spanish)

Currently translated at 87.3% (2111 of 2418 strings)

Translation: PeerTube/angular
Translate-URL: https://weblate.framasoft.org/projects/peertube/angular/es/

Commit:
Translated using Weblate (Spanish)

Currently translated at 100.0% (274 of 274 strings)

Translation: PeerTube/server
Translate-URL: https://weblate.framasoft.org/projects/peertube/server/es/

Commit:
Translated using Weblate (Spanish)

Currently translated at 100.0% (274 of 274 strings)

Translation: PeerTube/server
Translate-URL: https://weblate.framasoft.org/projects/peertube/server/es/

Commit:
Translated using Weblate (Spanish)

Currently translated at 100.0% (144 of 144 strings)

Translation: PeerTube/player
Translate-URL: https://weblate.framasoft.org/projects/peertube/player/es/

Alterak committed:
Translated using Weblate (Russian)

Currently translated at 98.7% (2388 of 2418 strings)

Translation: PeerTube/angular
Translate-URL: https://weblate.framasoft.org/projects/peertube/angular/ru/

Commit:
Translated using Weblate (Spanish)

Currently translated at 87.3% (2111 of 2418 strings)

Translation: PeerTube/angular
Translate-URL: https://weblate.framasoft.org/projects/peertube/angular/es/

Commit:
Translated using Weblate (Galician)

Currently translated at 100.0% (2418 of 2418 strings)

Translation: PeerTube/angular
Translate-URL: https://weblate.framasoft.org/projects/peertube/angular/gl/

s8321414 committed:
Translated using Weblate (Chinese (Traditional))

Currently translated at 100.0% (2418 of 2418 strings)

Translation: PeerTube/angular
Translate-URL: https://weblate.framasoft.org/projects/peertube/angular/zh_Hant/

milotype committed:
Translated using Weblate (Croatian)

Currently translated at 98.1% (2374 of 2418 strings)

Translation: PeerTube/angular
Translate-URL: https://weblate.framasoft.org/projects/peertube/angular/hr/

milotype committed:
Translated using Weblate (Croatian)

Currently translated at 98.2% (2375 of 2418 strings)

Translation: PeerTube/angular
Translate-URL: https://weblate.framasoft.org/projects/peertube/angular/hr/

milotype committed:
Translated using Weblate (Croatian)

Currently translated at 98.3% (2377 of 2418 strings)

Translation: PeerTube/angular
Translate-URL: https://weblate.framasoft.org/projects/peertube/angular/hr/

milotype committed:
Translated using Weblate (Croatian)

Currently translated at 98.5% (2383 of 2418 strings)

Translation: PeerTube/angular
Translate-URL: https://weblate.framasoft.org/projects/peertube/angular/hr/

milotype committed:
Translated using Weblate (Croatian)

Currently translated at 98.5% (2383 of 2418 strings)

Translation: PeerTube/angular
Translate-URL: https://weblate.framasoft.org/projects/peertube/angular/hr/

milotype committed:
Translated using Weblate (Croatian)

Currently translated at 100.0% (143 of 143 strings)

Translation: PeerTube/player
Translate-URL: https://weblate.framasoft.org/projects/peertube/player/hr/

milotype committed:
Translated using Weblate (Croatian)

Currently translated at 98.5% (2383 of 2418 strings)

Translation: PeerTube/angular
Translate-URL: https://weblate.framasoft.org/projects/peertube/angular/hr/

milotype committed:
Translated using Weblate (Croatian)

Currently translated at 98.6% (2385 of 2418 strings)

Translation: PeerTube/angular
Translate-URL: https://weblate.framasoft.org/projects/peertube/angular/hr/

milotype committed:
Translated using Weblate (Croatian)

Currently translated at 98.6% (2386 of 2418 strings)

Translation: PeerTube/angular
Translate-URL: https://weblate.framasoft.org/projects/peertube/angular/hr/

paymanzafar1 committed:
Translated using Weblate (Persian)

Currently translated at 87.6% (2120 of 2418 strings)

Translation: PeerTube/angular
Translate-URL: https://weblate.framasoft.org/projects/peertube/angular/fa/

Chocobozzz committed:
Integrate transcription in PeerTube

milotype committed:
Translated using Weblate (Croatian)

Currently translated at 99.4% (2404 of 2418 strings)

Translation: PeerTube/angular
Translate-URL: https://weblate.framasoft.org/projects/peertube/angular/hr/

Chirunos committed:
Translated using Weblate (Portuguese (Brazil))

Currently translated at 82.9% (2006 of 2418 strings)

Translation: PeerTube/angular
Translate-URL: https://weblate.framasoft.org/projects/peertube/angular/pt_BR/

Chocobozzz committed:
Fix short uuid use in custom playlist markup

Commit:
Translated using Weblate (Slovak)

Currently translated at 100.0% (98 of 98 strings)

Translation: PeerTube/player
Translate-URL: https://weblate.framasoft.org/projects/peertube/player/sk/

Commit:
Translated using Weblate (Slovak)

Currently translated at 91.7% (2218 of 2418 strings)

Translation: PeerTube/angular
Translate-URL: https://weblate.framasoft.org/projects/peertube/angular/sk/

lutangar commented on apps/peertube-runner/src/server/process/shared/process-transcription.ts:
```diff
@@ -0,0 +1,79 @@
+
+    const transcriptFile = await transcriber.transcribe({
+      mediaFilePath: inputPath,
+      model: config.modelPath
```
custom constructor to use or ad

Chocobozzz replied:
I'm not sure to understand your comment

lutangar replied:
Sorry, I left this comment in a hurry  as a side note to myself.

I tried to implement a custom constructor for these use cases in either `TranscriptionModel` (or `WhisperBuiltinModel` since this implementation is tied to **Whisper** as it is) but I failed to do so... in fact this should be achievable with the default constructor :thinking: 

But current constructor may lead to an invalid state (with a path which doesn't exists) and I'm not sure how to deal with this since there is `sync/async` dilemma and since is no such thing as async constructor...

lutangar replied:
*Btw the invalid object state is also possible with the `TranscriptFile` default constructor*

lutangar commented on apps/peertube-runner/src/server/process/shared/process-transcription.ts:
```diff
@@ -0,0 +1,79 @@
+      payload: successBody
+    })
+  } finally {
+    if (inputPath) await remove(inputPath)
```
Transcriber should probably handle  a part of the cleanup process when destroyed.

Chocobozzz replied:
I don't think so, the input file and output files can have a life after the transcriber is destroyed. We just want it to generate the files, that's all. I would agree to remove the output directory in the transcriber with a  `transcriber.guessLanguage` method for example

lutangar replied:
> I don't think so, the input file and output files can have a life after the transcriber is destroyed.

Agreed

> I would agree to remove the output directory in the transcriber with a transcriber.guessLanguage method for example

I don't quite get this part tho'

lutangar commented on config/default.yaml:
```diff
@@ -716,6 +716,34 @@ video_studio:
+
+  # Or specify the model path:
+  #  * PyTorch model file path for 'openai-whisper'
+  #  * CTranslate2 Whisper model directory path for 'openai-whisper'
```
```suggestion
  #  * CTranslate2 Whisper model directory path for 'whisper-ctranslate2'
```

Chocobozzz replied:
:+1:

lutangar commented on config/production.yaml.example:
```diff
@@ -726,6 +726,34 @@ video_studio:
+
+  # Or specify the model path:
+  #  * PyTorch model file path for 'openai-whisper'
+  #  * CTranslate2 Whisper model directory path for 'openai-whisper'
```
```suggestion
  #  * CTranslate2 Whisper model directory path for 'whisper-ctranslate2'
```

lutangar commented on packages/jiwer/requirements.txt:
The requirements file should be considered the standard to describe the Python dependencies.
This file path can be given as an argument to pip from the CLI. Might not be the best place for it tho'.

lutangar commented on packages/transcription-devtools/README.md:
````diff
@@ -0,0 +1,64 @@
+```
+
+## Jiwer
+-----
````
To remove

lutangar commented on packages/transcription-devtools/README.md:
````diff
@@ -0,0 +1,64 @@
+```
+
+Resources
+---------
````
```suggestion
## Resources
```

Consistent naming.

lutangar commented on packages/transcription/src/abstract-transcriber.ts:
```diff
@@ -59,11 +67,25 @@ export abstract class AbstractTranscriber {
     return model.format === 'PyTorch'
   }
 
+  getEngine () {
+    return this.engine
+  }
```
to comment

Chocobozzz replied:
Why?

lutangar replied:
Would it make sense to use Javascript `get` operator here ?
https://developer.mozilla.org/fr/docs/Web/JavaScript/Reference/Functions/get

lutangar commented on packages/transcription/src/whisper/transcriber/ctranslate2-transcriber.ts:
```diff
@@ -27,23 +31,33 @@ export class Ctranslate2Transcriber extends OpenaiTranscriber {
+    return model.format === 'CTranslate2'
+  }
+
+  async install (directory: string) {
```
:heart:

lutangar commented on packages/transcription/src/whisper/transcriber/openai-transcriber.ts:
```diff
@@ -1,31 +1,30 @@
   async transcribe ({
     mediaFilePath,
-    model = new WhisperBuiltinModel('tiny'),
+    model,
```
I love sane defaults !

Chocobozzz replied:
We prefer to force the caller to choose the model, as it's a very important choice depending on the context

lutangar commented on packages/transcription/src/whisper/transcriber/openai-transcriber.ts:
```diff
@@ -1,31 +1,30 @@
     this.assertLanguageDetectionAvailable(language)
 
-    const $$ = $({ verbose: process.env.NODE_ENV !== 'production' })
+    const $$ = $({ verbose: process.env.NODE_ENV !== 'production', env: this.getExecEnv() })
```
:+1:

lutangar commented on packages/transcription/src/whisper/transcriber/openai-transcriber.ts:
```diff
@@ -34,29 +33,45 @@ export class OpenaiTranscriber extends AbstractTranscriber {
 
-  async getDetectedLanguage (mediaFilePath: string) {
-    const { language } = await this.readJsonTranscriptFile(mediaFilePath)
+  // ---------------------------------------------------------------------------
```
I thought you wanted to group protected methods together and then I saw `getExecEnv` outside of this "block comment".
Although I like the idea of grouping methods by visibility (public /protected),  the "block comment" is hard to enforce programmatically and might appears a tad too subjective to other developers.

Chocobozzz replied:
I think I grouped methods by scope:

 * transcribe method
 * transcribed file methods
 * install and execa methods

lutangar replied:
I'd say it underlines the remark on subjectivity in my previous comment ;)

lutangar commented on server/core/initializers/config.ts:
```diff
@@ -515,6 +516,16 @@ const CONFIG = {
   },
+  VIDEO_TRANSCRIPTION: {
+    get ENABLED () { return config.get<boolean>('video_transcription.enabled') },
+    get ENGINE () { return config.get<TranscriptionEngineName>('video_transcription.engine') },
```
engine ~ binary ~ transcriber :thinking:

Chocobozzz replied:
I think we can rename `binary` to `command`.

Chocobozzz committed:
Support Service AP actors

lutangar reviewed (commented):
Hey @Chocobozzz , since I made a few comments already I might as well submit them right now. But I'll dig deeper tomorrow and add some more comments...

Chocobozzz committed:
Fix detecting account actor

Chocobozzz committed:
Fix channel update federation

Chocobozzz committed:
Fix bulk finder

Chocobozzz committed:
Detect internal link on plugin pages

Chocobozzz committed:
Prevent hotkeys in contenteditable element

Chocobozzz committed:
Fix line typo

Chocobozzz committed:
Fix lint

Chocobozzz committed:
Update translations

Chocobozzz committed:
Fix internal link detection

Links may also use mailto: or other schemes

Chocobozzz committed:
Fix internal link detection, again

Chocobozzz committed:
Update contributors list

Chocobozzz committed:
Fix trending sort label

Chocobozzz committed:
Highlight lives when listing videos

kontrollanten committed:
feat(plugins/filter): add req to api.video.get.result

Commit:
Translated using Weblate (Japanese)

Currently translated at 98.3% (2378 of 2418 strings)

Translation: PeerTube/angular
Translate-URL: https://weblate.framasoft.org/projects/peertube/angular/ja/

Goudarz committed:
Translated using Weblate (Persian)

Currently translated at 100.0% (145 of 145 strings)

Translation: PeerTube/player
Translate-URL: https://weblate.framasoft.org/projects/peertube/player/fa/

Chocobozzz committed:
Better onRootClick implementation

Thanks John

ujdhesa committed:
Translated using Weblate (Albanian)

Currently translated at 99.3% (144 of 145 strings)

Translation: PeerTube/player
Translate-URL: https://weblate.framasoft.org/projects/peertube/player/sq/

Chocobozzz committed:
Owner must not be able to approve its own comment

Chocobozzz committed:
Don't display divider on empty block

Chocobozzz committed:
Fix downloading private videos in admin

Chocobozzz committed:
href can be null

Chocobozzz committed:
Update server dependencies

Chocobozzz committed:
Update apps dependencies

Chocobozzz committed:
Fix lint and tests

Chocobozzz committed:
Fix loading actor involved in video

Chocobozzz committed:
Fix legacy upload req timeout

Chocobozzz committed:
Fix build

Commit:
Translated using Weblate (Ukrainian)

Currently translated at 89.7% (2169 of 2418 strings)

Translation: PeerTube/angular
Translate-URL: https://weblate.framasoft.org/projects/peertube/angular/uk/

Chocobozzz committed:
Update translations

Chocobozzz committed:
Fix lint

Chocobozzz committed:
Update client dependencies

lutangar commented on server/core/lib/video-captions.ts:
```diff
@@ -22,5 +40,122 @@ export async function createLocalCaption (options: {
+      const transcriptFile = await transcriber.transcribe({
+        mediaFilePath: videoInputPath,
+
+        model: CONFIG.VIDEO_TRANSCRIPTION.MODEL_PATH
```
Might be another usecase for the previsouly mentionned custom constructor.

lutangar commented on support/doc/development/tests.md:
````diff
@@ -36,11 +36,17 @@ Ensure you also have these commands:
 Otherwise, install the packages. On Debian-based systems (like Debian, Ubuntu or Mint):
 ```bash
 sudo apt-get install parallel libimage-exiftool-perl
+sudo pip install whisper-ctranslate2 openai-whisper
````
Mentions of these dependencies might be better gathered somewhere.

Chocobozzz replied:
They are only required for tests

lutangar commented on packages/transcription-devtools/package.json:
```diff
@@ -12,9 +12,9 @@
   "type": "module",
   "devDependencies": {},
   "scripts": {
-    "preinstall": "pip install -r requirements.txt",
```
While I undertstand `preinstall` script might not be the best place for this, I still think  the `requirements.txt` file as some part to play.

Chocobozzz committed:
Upgrade to angular 18 & vite

lutangar commented on scripts/ci.sh:
```diff
@@ -131,6 +131,8 @@ elif [ "$1" = "api-5" ]; then
 
     MOCHA_PARALLEL=true runJSTest "$1" $((2*$speedFactor)) $transcodingFiles $runnersFiles
 elif [ "$1" = "external-plugins" ]; then
+    pip install whisper-ctranslate2 openai-whisper jiwer
```
This list of dependencies should be kept somewhere in the package directory instead.
(if it's not in the `requirements.txt` file at least in a script command)

lutangar commented on scripts/ci.sh:
```diff
@@ -147,12 +149,13 @@ elif [ "$1" = "lint" ]; then
     ( cd client && npm run lint )
 elif [ "$1" = "transcription" ]; then
-    npm run preinstall --workspace=@peertube/peertube-transcription --workspace=@peertube/peertube-jiwer
+    pip install whisper-ctranslate2 openai-whisper jiwer
```
> see comment above

Chocobozzz committed:
Fix lint

Chocobozzz committed:
Remove bundlewatch

CI fails, our projects generates too many chunks unfortunately

Chocobozzz committed:
Add views tag to middlewares too

Chocobozzz committed:
Fix E2E tests

Chocobozzz committed:
Add server restart test

Chocobozzz committed:
Fix lint

Chocobozzz committed:
Metadata to know if the caption is auto generated

Chocobozzz committed:
Runner can choose job type

Chocobozzz committed:
Remove verbose option from transcription

Can be specified on-demand using NODE_DEBUG=execa env variable

Chocobozzz commented:
Merged manually in upstream develop branch: https://github.com/Chocobozzz/PeerTube/commit/1bfb791e0539df54d1d007683719dcb883870e1d

Thanks again!
